flexmeasures/data/models/planning/shiftable_load.py

```diff
@@ -0,0 +1,221 @@
+
+        optimization_sense: objective of the scheduler, to maximize or minimize.
+        time_restrictions: time periods in which the load cannot be schedule to.
+        load_type: Inflexible, Breakable or Shiftable.
+
+        :returns:               The computed schedule.
+        """
+
+        if not self.config_deserialized:
+            self.deserialize_config()
+
+        start = self.start.astimezone(pytz.utc)
+        end = self.end.astimezone(pytz.utc)
+        resolution = self.resolution
+        belief_time = self.belief_time
+        sensor = self.sensor
+
+        consumption_price_sensor: Sensor = self.flex_model.get(
+            "consumption_price_sensor"
+        )
+        duration: timedelta = self.flex_model.get("duration")
+        power = self.flex_model.get("power")
+        optimization_sense = self.flex_model.get("optimization_sense")
+        load_type: LoadType = self.flex_model.get("load_type")
+        time_restrictions = self.flex_model.get("time_restrictions")
+
+        # get cost data
+        cost = consumption_price_sensor.search_beliefs(
+            event_starts_after=start,
+            event_ends_before=end,
+            resolution=resolution,
+            one_deterministic_belief_per_event=True,
+            beliefs_before=belief_time,
+        )
+        cost = simplify_index(cost)
+
+        # create an empty schedule
+        schedule = pd.Series(
+            index=pd.date_range(
+                start,
+                end,
+                freq=sensor.event_resolution,
+                inclusive="left",
+                name="event_start",
+            ),
+            data=0,
+            name="event_value",
+        )
+
+        # optimize schedule for tomorrow. We can fill len(schedule) rows, at most
+        rows_to_fill = min(
+            ceil(duration / consumption_price_sensor.event_resolution), len(schedule)
+        )
+
+        # convert power to energy using the resolution of the sensor.
+        # e.g. resolution=15min, power=1kW -> energy=250W
+        energy = power * consumption_price_sensor.event_resolution / timedelta(hours=1)
+
+        if rows_to_fill > len(schedule):
+            raise ValueError(
+                f"Duration of the period exceeds the schedule window. The resulting schedule will be trimmed to fit the planning window ({start}, {end})."
+            )
+
+        # check if the time_restrictions allow for a load of the duration provided
+        if load_type in [LoadType.INFLEXIBLE, LoadType.SHIFTABLE]:
+            # get start time instants that are not feasible, i.e. some time during the ON period goes through
+            # a time restriction interval
+            time_restrictions = (
+                time_restrictions.rolling(duration).max().shift(-rows_to_fill + 1)
+            )
+            time_restrictions = (time_restrictions == 1) | time_restrictions.isna()
+
+            if time_restrictions.sum() == len(time_restrictions):
```

Comment:
> This tests if all slots are blocked, right?

yes, but I'm already taking into account the duration of the load. 

> We should test if there is no window of False values with a length >= the needed duration.

We are already testing that, let me explain:

The previous step:

```python
# get start time instants that are not feasible, i.e. some time during the ON period goes through
# a time restriction interval
time_restrictions = (
    time_restrictions.rolling(duration).max().shift(-rows_to_fill + 1)
)
time_restrictions = (time_restrictions == 1) | time_restrictions.isna()
```

is making sure to set to false the time steps that don't allow to fit a block of length `duration`. More technically, what I'm doing here is applying an [erosion](https://en.wikipedia.org/wiki/Dilation_(morphology)) of the time_restrictions array with a block of length `duration`. 

Then, the condition `if time_restrictions.sum() == len(time_restrictions):`, makes sure that at least we have a spot to place the load.

For example:

```python
time_restriction = [1 0 0 1 1 1 0 0 1 0]

# applying a dilation with duration = 2
time_restriction = [1 0 1 1 1 1 0 1 1 1]
```

We can only fit a block of `duration = 2` in the positions 1 and 6. `sum(time_restrictions) == 8`, while the `len(time_restriction) == 10`, which means we have `10-8=2` positions.

Reply:
Your example would work for BREAKABLE, but this is happening in the block for the other types.

Reply:
For the other two types, the `time_restrictions.rolling(duration).max().shift(-rows_to_fill + 1)` would to set to False those periods where we can't start the load, as we would encounter some other time restriction during the activation of the load. After that, we need to check if there is at least one place to start the load.

Reply:
In the case of `BREAKABLE`, we just need to ensure that there are `duration` blocks available.

Reply:
Ah yes, I misread your example. I get it now.

I believe the more straightforward (and readable) way to test here is to explicitly check for at least one valid starting time being present (and also use those words in a comment):

`(~time_restrictions).sum() > 0` or (just guessing this one) `~time_restrictions.any()`

Reply:
Perfect! 

I think the former is more straightforward, but instead of `(~time_restrictions).sum() > 0`, the condition should be  `(~time_restrictions).sum() == 0`

Reply:
I actually like the approach with `any()` more as it communicates in natural language what we need: Any valid starting time.

`(~time_restrictions).sum() == 0` is the condition for raising right? So in the other method, we should raise if `time_restrictions.all()`. Or with my proposed renaming: `start_time_restrictions.all()`.